src/hotspot/share/opto/type.hpp

```diff
@@ -551,7 +556,12 @@ class TypeD : public Type {
 
 class TypeInteger : public Type {
 protected:
-  TypeInteger(TYPES t, int w) : Type(t), _widen(w) {}
+  TypeInteger(TYPES t, int w, bool dual) : Type(t), _is_dual(dual), _widen(w) {}
+
+  // Denote that a set is a dual set.
+  // Dual sets are only used to compute the join of 2 sets, and not used
+  // outside.
+  const bool _is_dual;
```

Comment:
Do you have a clear definition of what a dual is somewhere?

Reply:
Tbh I'm not entirely sure what does "dual" serve other than: it is to compute the join of 2 sets. Looking at the code suggests that a `Type` is somehow a symmetric space with all normal types on one side and the "dual" types on the other. This really confuses me and I cannot figure out what exactly a dual type represents.

Reply:
My understanding is "join" means "union", "meet" means "intersection", and "dual" means "complement".

Reply:
"Complement" may not be quite right, or at least I don't see how it applies to TypeF and TypeD, whose xdual() functions are "self-symmetric" and do nothing except return the original type unchanged..

Reply:
https://en.wikipedia.org/wiki/Duality_(order_theory)

Reply:
If _lo <= x <= _hi, then I believe the dual is _hi <= x <= _lo 

If dual is really only needed for join, then it seems like we could remove the concept of dual and just implement join.

Reply:
@dean-long Thanks, that is really helpful. IIUC, the duality here refers to the set of all `TypeInt` with a set `a` considered higher than `b` if `a` is a subset of `b`. This leads to our notion of bottom type being the universe set and top type being the empty set. It still does not make sense for the concept of a dual `TypeInt`, though, since the concept of duality applies to the set of `TypeInt`, not the `TypeInt`s themselves.

> My understanding is "join" means "union", "meet" means "intersection", and "dual" means "complement".

You got it backward, "join" means intersection and "meet" means union.

Reply:
If you want to understand full details of how a (symmetrical) type lattice with duals supports a unified model for many different type flow analysis algorithms you can read up on it in Nielsen, Nielsen and Hankin's book Principles of Program Analysis. If it is new to you then a more simplified account of the use of (unqualified) TOP and BOTTOM types in type flow analysis can be found in Muchnick's book Advanced Compiler Design and Implementation. Note that Cliff Click goes against conventional mathematical terminology in making BOTTOM a universal type and TOP an empty (unrealizable) type.

One detail that may not be obvious is that the sub-lattice for int and long sorts includes the hierarchy of single, continuous intervals. Individual integral values (on the lattice centre line) are modelled as singleton ranges i.e. [a,a]. Given the large cardinality of the set of continuous intervals this makes it necessary to place a bound on any fixed point iterations that widen interval ranges. The iteration is killed by widening to the maximum range (this is what Cliff refers to in the code as a 'death march').

Reply:
@adinn Thanks a lot for your direction, it is really interesting and took me a while to read through. Although, I think that in practice, currently C2 only use `dual` to compute the join of 2 types, which is rather confusing.

Reply:
I'm OK with this clever "lazy dual" optimization using _is_dual that get reversed instead of doing an actual computation, but how can we make sure it is only used with join() and not accidentally used with other operations?